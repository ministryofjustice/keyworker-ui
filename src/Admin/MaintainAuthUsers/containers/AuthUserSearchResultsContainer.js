/* eslint-disable radix */
import React, { Component } from 'react'
import PropTypes from 'prop-types'
import { connect } from 'react-redux'
import axios from 'axios/index'
import qs from 'query-string'
import ReactRouterPropTypes from 'react-router-prop-types'
import { withRouter } from 'react-router'
import { setError, resetError, setLoaded, setMaintainAuthUsersList } from '../../../redux/actions/index'
import { authUserListType, errorType } from '../../../types'
import Page from '../../../Components/Page'
import AuthUserSearchResults from '../components/AuthUserSearchResults'
import validateSearch from './AuthUserSearchValidation'

class AuthUserSearchResultsContainer extends Component {
  constructor(props) {
    super()
    this.handleSearch = this.handleSearch.bind(this)
    props.resetErrorDispatch()
  }

  async componentDidMount() {
    await this.performSearch()
  }

  async componentDidUpdate(prevProps) {
    const { location } = this.props

    if (prevProps.location.search !== location.search) {
      await this.performSearch()
    }
  }

  handleChange = event => {
    const { name, value } = event.target
    this.setState({ [name]: value })
  }

  async performSearch() {
    const {
      location: { search },
      userListDispatch,
      setLoadedDispatch,
      resetErrorDispatch,
      handleError,
    } = this.props

    const { user } = qs.parse(search)
    if (!validateSearch(user, handleError)) {
      userListDispatch([])
      return
    }

    setLoadedDispatch(false)
    try {
      const users = await axios.get('/api/auth-user-search', {
        params: {
          nameFilter: user,
        },
      })
      userListDispatch(users.data)
      resetErrorDispatch()
    } catch (error) {
      userListDispatch([])
      handleError(error)
    }

    setLoadedDispatch(true)
  }

  async handleSearch(event) {
    const { history } = this.props
    const userQuery = qs.stringify(this.state)

    event.preventDefault()
    history.push({ pathname: '/admin-utilities/maintain-auth-users/search-results', search: userQuery })
  }

  render() {
    const {
      location: { search },
      userList,
      error,
    } = this.props
    const { user } = qs.parse(search)
    return (
      <Page title="Search for auth user results" alwaysRender>
        <AuthUserSearchResults
          handleChange={this.handleChange}
          handleSearch={this.handleSearch}
          user={user}
          userList={userList}
          error={error}
        />
      </Page>
    )
  }
}

AuthUserSearchResultsContainer.propTypes = {
  location: ReactRouterPropTypes.location.isRequired,
  history: ReactRouterPropTypes.history.isRequired,
  resetErrorDispatch: PropTypes.func.isRequired,
  setLoadedDispatch: PropTypes.func.isRequired,
  userListDispatch: PropTypes.func.isRequired,
  handleError: PropTypes.func.isRequired,
  userList: authUserListType.isRequired,
  error: errorType.isRequired,
}

const mapStateToProps = state => ({
  error: state.app.error,
  config: state.app.config,
  userList: state.maintainAuthUsers.userList,
  loaded: state.app.loaded,
})

const mapDispatchToProps = dispatch => ({
  userListDispatch: list => dispatch(setMaintainAuthUsersList(list)),
  setErrorDispatch: error => dispatch(setError(error)),
  resetErrorDispatch: () => dispatch(resetError()),
  setLoadedDispatch: status => dispatch(setLoaded(status)),
})

export default withRouter(
  connect(
    mapStateToProps,
    mapDispatchToProps
  )(AuthUserSearchResultsContainer)
)
