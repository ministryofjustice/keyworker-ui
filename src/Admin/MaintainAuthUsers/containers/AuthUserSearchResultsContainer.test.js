import React from 'react'
import { Provider } from 'react-redux'
import { MemoryRouter } from 'react-router-dom'
import { shallow, mount } from 'enzyme'

import AuthUserSearchResultsContainer from './AuthUserSearchResultsContainer'

describe('Auth search results container', () => {
  it('should render correctly without existing location', () => {
    const wrapper = shallow(<AuthUserSearchResultsContainer location={{ search: {} }} />)
    expect(wrapper).toMatchSnapshot()
  })

  it('should render correctly with search location', () => {
    const wrapper = shallow(<AuthUserSearchResultsContainer location={{ search: { user: 'auser' } }} />)
    expect(wrapper).toMatchSnapshot()
  })

  describe('handle functions', () => {
    const event = { target: { name: 'user', value: 'usersearched' }, preventDefault: jest.fn() }
    const store = { subscribe: jest.fn(), dispatch: jest.fn(), getState: jest.fn(), setState: jest.fn() }
    store.getState.mockReturnValue({ app: { error: '', loaded: true }, maintainAuthUsers: { userList: [] } })

    const wrapper = mount(
      <Provider store={store}>
        <MemoryRouter>
          <AuthUserSearchResultsContainer handleError={jest.fn()} />
        </MemoryRouter>
      </Provider>
    )
    wrapper.find('input#user').simulate('change', event)

    it('should set the user input on the state', () => {
      expect(wrapper.find('AuthUserSearchResultsContainer').state().user).toEqual('usersearched')
    })

    it('should set history when form submitted', () => {
      wrapper.find('form').simulate('submit')

      const {
        location: { pathname, search },
      } = wrapper.find('AuthUserSearchResultsContainer').props()

      expect(pathname).toEqual('/admin-utilities/maintain-auth-users/search-results')
      expect(search).toEqual('?user=usersearched')
    })

    it('should prevent default on the form submission', () => {
      const submitEvent = { target: { value: 'search' }, preventDefault: jest.fn() }
      wrapper.find('form').simulate('submit', submitEvent)

      expect(submitEvent.preventDefault).toBeCalled()
    })
  })
})
