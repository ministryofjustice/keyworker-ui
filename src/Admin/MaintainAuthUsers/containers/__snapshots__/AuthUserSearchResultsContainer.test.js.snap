// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Auth search results container should render correctly with search location 1`] = `ShallowWrapper {}`;

exports[`Auth search results container should render correctly without existing location 1`] = `ShallowWrapper {}`;
